getting_started/step_by_step/resources.rst

```diff
@@ -158,12 +158,173 @@ Freeing resources
+        health = p_health
+        sub_resource = p_sub_resource
+        str_array = p_str_array
+
+    # bot.gd
+    extends KinematicBody
+    export(Resource) var stats
+
+    const BotStats = preload("res://bot_stats.gd")
+
+    func _init():
+        if not stats:
+            stats = BotStats.new(10)
+    
+    func _ready():
+        print(stats.health) # prints '10'
+  ..code-tab:: csharp
+        using Godot;
+
+        public class Bot : KinematicBody
+        {
+            [Export]
+            public BotStats stats;
+
+            public override void _Init()
+            {
+                if (!stats)
+                {
+                    stats = (ResourceLoader::load("BotStats.cs") as BotStats).new(10);
+                }
+            }
+
+            public override void _Ready()
+            {
+                Godot.print(stats.health);
+            }
+        }
+
+.. note::
+
+   Resource scripts are similar to Unity's ScriptableObjects. However, unlike in
+   Unity, there is no need to implement IMGUI logic to enable users to create
+   ScriptableObject files from the Editor. The Inspector provides built-in 
+   support for custom resources. If desired, users can even design their own
+   Control-based tool scripts and combine them with an
+   :ref:`EditorPlugin <class_EditorPlugin>` to create custom visualizations
+   and editors for their data.
+
+   Unreal Engine 4's DataTables and CurveTables are also easy to recreate with
+   Resource scripts. DataTables are a String mapped to a custom struct, similar
+   to a Dictionary mapping a String to a secondary custom Resource script.
+
+    .. tabs::
+      .. code-tab:: gdscript GDScript
+
+        extends Resource
+
+        const BotStats = preload("bot_stats.gd") 
+
+        func _init():
+            var dict = {
+                "GodotBot": BotStats.new(10), # creates instance with 10 health
+                "DifferentBot": BotStats.new(20) # a different one with 20 health
+            }
+            print(dict)
+      .. code-tab:: csharp
+        using Godot;
+
+        public class BotStatsTable : Resource
+        {
+            private Godot.Dictionary<String, BotStats> stats = new Godot.Dictionary<String, BotStats>();
+
+            public override void _Init()
+            {
+                stats["GodotBot"] = new BotStats(10);
+                stats["DifferentBot"] = new BotStats(20);
+                Godot.print(stats);
+            }
+        }
+    
+    CurveTables are the same thing except mapped to an Array of float values
+    or a :ref:`Curve <class_Curve>`/:ref:`Curve2D <class_Curve2D>` resource object.
+
+.. warning::
+
+    With GDScript, it only supports Resource classes at the top level. Defining
+    a Resource subclass, i.e. an inner class, within a GDScript file will have no 
+    effect. If one attempts to save an instance of a Resource subclass, none of 
+    the scripted properties will serialize to the saved file.
```

Comment:
This sounds like a bug to me -- I've saved `ShaderMaterial`s with `Shader` resources inside them before. I don't see why this should be any different.

Reply:
You can have sub resources within scripts, but if you try to define an entirely new type of Resource *as a subclass within another script*, then Godot isn't going to identify the path for the script on that resource as "path to the file that has the Resource subclass AND somehow mark that you need to assign the constant value of the subclass Script resource as the script value for the created \*.tres. You know, something like "res://file.gd#MyResource" or what-have-you.

Reply:
Like, if I make a Resource object and set it's script to be not another script file, but a Resource class that's a subclass of a script file (where the script file could derive ANYTHING), it won't start serializing those script properties properly or anything.

Reply:
Thing is, the way this issue is being described in this warning makes it come off as a bug. It sounds like something Godot *could* do, and that it probably should do, but for one reason or another just... doesn't do.

Maybe try to explain why this is the case, at least?

Reply:
Sure, sounds good.